### PR TITLE
Hashcode calculation update

### PR DESCRIPTION
**Pull request by buyaa-n**

### CODE REVIEW

File: src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs [hunk 1]
```diff
@@ -182,18 +183,9 @@ public static int GetHashCode(string name)
...
-                    {
-                        char ch = name[i];
-                        if (ch == ':') break;
-                        hashCode += (hashCode << 7) ^ ch;
-                    }
-                    hashCode -= hashCode >> 17;
-                    hashCode -= hashCode >> 11;
-                    hashCode -= hashCode >> 5;
-                }
+                string[] splitted = name.Split(":");
```
Comment (danmosemsft) [inline src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs]: "splitted" -> "split"

File: src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs [hunk 2]
```diff
@@ -182,18 +183,9 @@ public static int GetHashCode(string name)
...
+                ReadOnlySpan<byte> bytes = MemoryMarshal.AsBytes(splitted[splitted.Length - 1].AsSpan());
```
Comment (EgorBo) [inline src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs]: the old code didn't allocate (`string.Split`).
```
name.AsSpan(name.LastIndexOf(':')) ?
```

Comment (buyaa-n) [reply src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs]: colon shouldn't included

Comment (krwq) [reply src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs]: @buyaa-n - this cannot allocate - this is being called hundreds of times

File: src/System.Private.Xml.Linq/src/System/Xml/Linq/XHashtable.cs [hunk 3]
```diff
@@ -397,22 +398,9 @@ private bool FindEntry(int hashCode, string key, int index, int count, ref int e
...
-                    {
-                        hashCode += (hashCode << 7) ^ key[i];
-                    }
-
-                // Mix up hash code a bit more and clear the sign bit.  This code was taken from NameTable.cs in System.Xml.
-                hashCode -= hashCode >> 17;
-                hashCode -= hashCode >> 11;
-                hashCode -= hashCode >> 5;
-                return hashCode & 0x7FFFFFFF;
+                ReadOnlySpan<byte> bytes = MemoryMarshal.AsBytes(new ReadOnlySpan<char>(key.ToCharArray(), index, count));
```
Comment (bartonjs) [inline src/System.Private.Xml.Linq/src/System/Xml/Linq/XHashtable.cs]:
```suggestion
                ReadOnlySpan<byte> bytes = MemoryMarshal.AsBytes(key.AsSpan(index, count));
```

The call to ToCharArray allocates, but that's not necessary.

File: src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs (hunk 2 above)
Comment (bartonjs) [inline src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs]: Hmm, isn't it always just `name.AsSpan(name.LastIndexOf(':') + 1)`?

If there's no colon, LastIndexOf returns -1, so +1 is 0 (the whole string).  If there is one, then it starts at the next character.

Comment (buyaa-n) [reply src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs]: yes, that is much better, thanks!

File: src/System.Private.Xml.Linq/src/System/Xml/Linq/XHashtable.cs (hunk 3 above)
Comment (danmosemsft) [reply src/System.Private.Xml.Linq/src/System/Xml/Linq/XHashtable.cs]: BTW, is it necessary to do `fixed` if you do `AsBytes`? Or this is all handled by JIT/runtime magic.

File: src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs (hunk 1 above)
Comment (bbartels) [reply src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs]:
```suggestion
                string[] split = name.Split(':');
```

File: src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs (hunk 2 above)
Comment (bbartels) [reply src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs]: Maybe wait until https://github.com/dotnet/corefx/issues/26528 is merged? Would not require allocating for split @krwq

Comment (stephentoub) [reply src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs]: There's no need to Split here at all.

File: src/System.Private.Xml.Linq/src/System/Xml/Linq/XHashtable.cs (hunk 3 above)
Comment (bartonjs) [reply src/System.Private.Xml.Linq/src/System/Xml/Linq/XHashtable.cs]: I believe "magic".  A `ref` is still pointing within the object data payload, so the GC should still see the target as alive.

**Comment by bartonjs** [comment]
@buyaa-n it looks like (as of 2daf8bc) the XHashtable version is still allocating (ToCharArray).

File: src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs [hunk 4]
```diff
@@ -182,18 +183,8 @@ public static int GetHashCode(string name)
...
-                        char ch = name[i];
-                        if (ch == ':') break;
-                        hashCode += (hashCode << 7) ^ ch;
-                    }
-                    hashCode -= hashCode >> 17;
-                    hashCode -= hashCode >> 11;
-                    hashCode -= hashCode >> 5;
-                }
+                ReadOnlySpan<byte> bytes = MemoryMarshal.AsBytes(name.AsSpan(name.LastIndexOf(':') + 1));
+                return Marvin.ComputeHash32(bytes, Marvin.DefaultSeed);
```
Comment (krwq) [inline src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs]: Supposedely those two functions don't exchange hash code? They are in different assemblies which means they have two copies of Marvin with possibly different seed

File: src/System.Private.Xml.Linq/src/System/Xml/Linq/XHashtable.cs [hunk 5]
```diff
@@ -397,22 +398,9 @@ private bool FindEntry(int hashCode, string key, int index, int count, ref int e
...
-                    {
-                        hashCode += (hashCode << 7) ^ key[i];
-                    }
-
-                // Mix up hash code a bit more and clear the sign bit.  This code was taken from NameTable.cs in System.Xml.
-                hashCode -= hashCode >> 17;
-                hashCode -= hashCode >> 11;
-                hashCode -= hashCode >> 5;
-                return hashCode & 0x7FFFFFFF;
+                ReadOnlySpan<byte> bytes = MemoryMarshal.AsBytes(key.AsSpan(index, index + count));
```
Comment (bartonjs) [inline src/System.Private.Xml.Linq/src/System/Xml/Linq/XHashtable.cs]: https://source.dot.net/#System.Private.CoreLib/shared/System/MemoryExtensions.Fast.cs,e3849b379849b4a4

The arguments to AsSpan are `start` and `length`, which should just be `index` and `count`... so adding index to count (to produce `end`) seems wrong.

Comment (buyaa-n) [reply src/System.Private.Xml.Linq/src/System/Xml/Linq/XHashtable.cs]: uh, I thought it is end index

File: src/System.Private.Xml.Linq/src/System.Private.Xml.Linq.csproj [hunk 6]
```diff
@@ -19,6 +19,9 @@
     <Compile Include="$(CommonPath)\CoreLib\System\Text\StringBuilderCache.cs">
       <Link>System\Text\StringBuilderCache.cs</Link>
     </Compile>
+    <Compile Include="$(CommonPath)\System\Marvin.cs">
+      <Link>System\Marvin.cs</Link>
```
Comment (GrabYourPitchforks) [inline src/System.Private.Xml.Linq/src/System.Private.Xml.Linq.csproj]: We don't need to pull in our own copy of `Marvin`. All APIs we need are already publicly accessible via the `string` or `HashCode` types.

File: src/System.Private.Xml.Linq/src/System/Xml/Linq/XHashtable.cs (hunk 5 above)
Comment (GrabYourPitchforks) [inline src/System.Private.Xml.Linq/src/System/Xml/Linq/XHashtable.cs]:
```cs
return string.GetHashCode(key.AsSpan(index, count));
```

File: src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs [hunk 7]
```diff
@@ -182,18 +183,8 @@ public static int GetHashCode(string name)
...
-                    {
-                        char ch = name[i];
-                        if (ch == ':') break;
-                        hashCode += (hashCode << 7) ^ ch;
-                    }
-                    hashCode -= hashCode >> 17;
-                    hashCode -= hashCode >> 11;
-                    hashCode -= hashCode >> 5;
-                }
+                ReadOnlySpan<byte> bytes = MemoryMarshal.AsBytes(name.AsSpan(name.LastIndexOf(':') + 1));
```
Comment (GrabYourPitchforks) [inline src/System.Private.Xml/src/System/Xml/Dom/XmlName.cs]:
```cs
hashCode = string.GetHashCode(name.AsSpan(name.LastIndexOf(':') + 1));
```

File: src/System.Private.Xml.Linq/src/System/Xml/Linq/XHashtable.cs (hunk 5 above)
Comment (krwq) [reply src/System.Private.Xml.Linq/src/System/Xml/Linq/XHashtable.cs]: I didn't know we added this :) perhaps it is possible to remove Marvin from Common now?

Comment (GrabYourPitchforks) [reply src/System.Private.Xml.Linq/src/System/Xml/Linq/XHashtable.cs]: This API was added in netcoreapp 3.0. So for things which are inbox or build as _netcoreapp_ or _uapaot_, this API should be accessible. I believe this particular project qualifies.

For things which are fully OOB and which are cross-compiled against _netstandard_ / _netfx_ / etc., the private implementation will have to be carried. So unfortunately we can't delete the private implementation just yet.

**Pull request merged**